cranelift/codegen/src/opts.rs

```diff
@@ -0,0 +1,309 @@
+                ctx.egraph.stats.node_subsume += 1;
+                // Merge in the unionfind so canonicalization still
+                // works, but take *only* the subsuming ID, and break
+                // now.
+                ctx.egraph.egraph.unionfind.union(union_id, new_id);
+                union_id = new_id;
+                break;
+            }
+            ctx.egraph.stats.node_union += 1;
+            let old_union_id = union_id;
+            union_id = ctx
+                .egraph
+                .egraph
+                .union(&ctx.egraph.node_ctx, union_id, new_id);
+            trace!(
+                " -> union eclass {} with {} to get {}",
+                new_id,
+                old_union_id,
+                union_id
+            );
+        }
+    }
+    ctx.egraph.subsume_ids.clear();
+    trace!(" -> optimize {} got {}", id, union_id);
+    ctx.egraph.rewrite_depth -= 1;
+    union_id
+}
+
+pub(crate) fn store_to_load<'a>(id: Id, egraph: &mut FuncEGraph<'a>) -> Id {
+    if let Some(load_key) = egraph.egraph.classes[id].get_node() {
+        if let Node::Load {
+            op: load_op,
+            ty: load_ty,
+            addr: load_addr,
+            mem_state: MemoryState::Store(store_inst),
+            ..
+        } = load_key.node::<NodeCtx>(&egraph.egraph.nodes)
+        {
+            trace!(" -> got load op for id {}: {:?}", id, load_op);
+            if let Some((store_ty, store_id)) = egraph.store_nodes.get(&store_inst) {
+                trace!(" -> got store id: {} ty: {}", store_id, store_ty);
+                if *store_ty == *load_ty {
+                    if let Some(store_key) = egraph.egraph.classes[*store_id].get_node() {
+                        if let Node::Inst {
+                            op: store_op,
+                            args: store_args,
+                            ..
+                        } = store_key.node::<NodeCtx>(&egraph.egraph.nodes)
+                        {
+                            trace!(
+                                "load id {} from store id {}: {:?}, {:?}",
+                                id,
+                                store_id,
+                                load_op,
+                                store_op
+                            );
+                            match (load_op, store_op) {
+                                (
+                                    InstructionImms::Load {
+                                        opcode: Opcode::Load,
+                                        offset: load_offset,
+                                        ..
+                                    },
+                                    InstructionImms::Store {
+                                        opcode: Opcode::Store,
+                                        offset: store_offset,
+                                        ..
+                                    },
+                                ) if *load_offset == *store_offset => {
```

Comment:
I golfed this a bit more, starting from your suggestions (thanks!); I combined the three conditionals (type, offset, address) into one as well. Much better now!

I'm still noodling over the idea of making this an ISLE rule instead; maybe eventually it will be better that way, when we have more infrastructure to support non-pure ops, but I think I want to approach that idea incrementally / as a followup.